lte/gateway/c/connection_tracker/EventTracker.cpp

```diff
@@ -0,0 +1,274 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <iostream>
+
+#include <libmnl/libmnl.h>
+#include <linux/netfilter/nfnetlink.h>
+#include <linux/netfilter/nfnetlink_conntrack.h>
+
+#include <linux/if_packet.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <net/if.h>
+#include <netinet/ether.h>
+#include <linux/ip.h>
+#include <memory>
+
+#include "EventTracker.h"
+
+#include "magma_logging.h"
+
+
+static int data_cb(const struct nlmsghdr* nlh, void* data);
+
+namespace magma {
+namespace lte {
+
+EventTracker::EventTracker(std::shared_ptr<PacketGenerator> pkt_gen)
+    : pkt_gen_(pkt_gen) {}
+
+int EventTracker::init_conntrack_event_loop(void) {
+  struct mnl_socket* nl;
+  char buf[MNL_SOCKET_BUFFER_SIZE];
+  int ret;
+
+  nl = mnl_socket_open(NETLINK_NETFILTER);
+  if (nl == NULL) {
+    perror("mnl_socket_open");
+    exit(EXIT_FAILURE);
+  }
+
+  if (mnl_socket_bind(
+          nl,
+          NF_NETLINK_CONNTRACK_NEW |
+              // NF_NETLINK_CONNTRACK_UPDATE |
+              NF_NETLINK_CONNTRACK_DESTROY,
+          MNL_SOCKET_AUTOPID) < 0) {
+    perror("mnl_socket_bind");
+    exit(EXIT_FAILURE);
+  }
+
+  while (1) {
+    ret = mnl_socket_recvfrom(nl, buf, sizeof(buf));
+    if (ret == -1) {
+      perror("mnl_socket_recvfrom");
+      exit(EXIT_FAILURE);
+    }
+    ret = mnl_cb_run(buf, ret, 0, 0, data_cb, (void*) pkt_gen_.get());
+    if (ret == -1) {
+      perror("mnl_cb_run");
+      exit(EXIT_FAILURE);
+    }
+  }
+
```

Comment:
netlink can loose data in case of heavy load. can you investigate how to avoid it and detect it if it is un avoidable.
We need to handle the case by polling connections from kernel table. It can be TODO for now. But we need mechanism to identify that we lost events from conntrack.

Reply:
Interesting, okay I'll research and have a task for it